File: android/app/src/main/java/com/hermes/client/ui/sessions/SearchText.kt
package com.hermes.client.ui.sessions

/**
 * Pure text helpers for search result rows (docs/DESIGN.md §5.2 搜索页). No Compose here so the
 * rules are unit-testable: where the query occurs in a string, and how a long snippet is cut so
 * the first occurrence stays visible.
 */

/** Every case-insensitive, non-overlapping occurrence of [query] in [text]; empty for a blank query. */
fun highlightRanges(text: String, query: String): List<IntRange> {
    val q = query.trim()
    if (q.isEmpty() || text.isEmpty()) return emptyList()
    val ranges = mutableListOf<IntRange>()
    var from = 0
    while (from <= text.length - q.length) {
        val at = text.indexOf(q, from, ignoreCase = true)
        if (at < 0) break
        ranges += at until at + q.length
        from = at + q.length
    }
    return ranges
}

/**
 * Cuts [raw] to a window around the first occurrence of [query] (±[context] chars, whitespace
 * collapsed) so the match is visible in a two-line row. Without a match the head of the text is
 * kept. The gateway's own snippet varies by search path (short for the CJK trigram path, 120 for
 * the LIKE fallback, ~330 for FTS5), so the client normalizes here.
 */
fun centerSnippet(raw: String?, query: String, context: Int = 40): String {
    val text = raw.orEmpty().replace(Regex("\\s+"), " ").trim()
    if (text.isEmpty()) return ""
    val q = query.trim()
    val at = if (q.isEmpty()) -1 else text.indexOf(q, ignoreCase = true)
    val start: Int
    val end: Int
    if (at < 0) {
        start = 0
        end = (2 * context + q.length).coerceAtMost(text.length)
    } else {
        start = (at - context).coerceAtLeast(0)
        end = (at + q.length + context).coerceAtMost(text.length)
    }
    val core = text.substring(start, end).trim()
    val prefix = if (start > 0 && !core.startsWith("…")) "…" else ""
    val suffix = if (end < text.length && !core.endsWith("…")) "…" else ""
    return "$prefix$core$suffix"
}

/** Whether [session] is a title match for [query]: title or project label contains it. */
fun titleMatches(title: String, projectLabel: String?, query: String): Boolean {
    val q = query.trim()
    if (q.isEmpty()) return false
    return title.contains(q, ignoreCase = true) || (projectLabel?.contains(q, ignoreCase = true) == true)
}

File: web/src/app/searchText.ts
// Search result text rules (Android ui/sessions/SearchText.kt): where the query occurs, and a
// snippet cut so the first occurrence stays visible in a two-line row.

/** Case-insensitive, non-overlapping occurrences as [start, end) pairs; empty for a blank query. */
export function highlightRanges(text: string, query: string): [number, number][] {
  const q = query.trim().toLowerCase();
  if (!q || !text) return [];
  const lower = text.toLowerCase();
  const out: [number, number][] = [];
  let from = 0;
  while (from <= lower.length - q.length) {
    const at = lower.indexOf(q, from);
    if (at < 0) break;
    out.push([at, at + q.length]);
    from = at + q.length;
  }
  return out;
}

/** A window of ±`context` characters around the first match, whitespace collapsed, with ellipses. */
export function centerSnippet(raw: string | null | undefined, query: string, context = 40): string {
  const text = (raw ?? "").replace(/\s+/g, " ").trim();
  if (!text) return "";
  const q = query.trim();
  const at = q ? text.toLowerCase().indexOf(q.toLowerCase()) : -1;
  const start = at < 0 ? 0 : Math.max(0, at - context);
  const end = at < 0 ? Math.min(text.length, 2 * context + q.length) : Math.min(text.length, at + q.length + context);
  const core = text.slice(start, end).trim();
  return `${start > 0 && !core.startsWith("…") ? "…" : ""}${core}${end < text.length && !core.endsWith("…") ? "…" : ""}`;
}

export function titleMatches(title: string, projectLabel: string | null, query: string): boolean {
  const q = query.trim().toLowerCase();
  if (!q) return false;
  return title.toLowerCase().includes(q) || (projectLabel?.toLowerCase().includes(q) ?? false);
}
